File: DotNetGotchas/CSharp/CopyingObjects/Copy/Person.cs
//Person.cs

using System;

namespace Copy
{
	public class Person
	{
		private int theAge;
		private Brain theBrain;

		public Person(int age, Brain aBrain)
		{
			theAge = age;
			theBrain = aBrain;
		}

		public Person(Person another)
		{
			theAge = another.theAge;
			theBrain = new Brain(another.theBrain);
			// You rely on the copy constructor of Brain
			//to make a good deep copy
		}

		public override string ToString()
		{
			return "This is person with age " +
					   theAge + " and " + 
                       theBrain;
		}

	}
}

File: DotNetGotchas/VBDN/CopyingObjects/Copy/Person.vb
'Person.vb
Public Class Person
    Private theAge As Integer
    Private theBrain As Brain

    Public Sub New(ByVal age As Integer, ByVal aBrain As Brain)
        theAge = age
        theBrain = aBrain
    End Sub

    Public Sub New(ByVal another As Person)
        theAge = another.theAge
        theBrain = New Brain(another.theBrain)
        ' You rely on the copy constructor of Brain
        ' to make a good deep copy
    End Sub

    Public Overrides Function ToString() As String
        Return "This is person with age " & _
            theAge & " and " & _
            theBrain.ToString()
    End Function
End Class
